src/stan/model/indexing/rvalue.hpp

```diff
@@ -27,168 +28,138 @@ namespace model {
- * vector.
+ * Return the result of indexing the specified Eigen matrix with a
+ * sequence consisting of one single index, returning a vector.
  *
- * Types:  row_vec[multiple] : rowvec
+ * Types:  mat[,single] : vec
  *
- * @tparam T Scalar type.
- * @tparam I Multi-index type.
- * @param[in] rv Eigen row vector.
- * @param[in] idx Index consisting of one multi-index.
+ * @tparam EigMat An eigen matrix
+ * @param[in] a Eigen matrix.
+ * @param[in] idx Index consisting of one uni-index.
  * @param[in] name String form of expression being evaluated.
  * @param[in] depth Depth of indexing dimension.
- * @return Result of indexing vector.
+ * @return Result of indexing matrix.
  */
-template <typename T, typename I>
-inline std::enable_if_t<!std::is_same<I, index_uni>::value,
-                        Eigen::Matrix<T, 1, Eigen::Dynamic> >
-rvalue(const Eigen::Matrix<T, 1, Eigen::Dynamic>& rv,
-       const cons_index_list<I, nil_index_list>& idx, const char* name = "ANON",
-       int depth = 0) {
-  int size = rvalue_index_size(idx.head_, rv.size());
-  Eigen::Matrix<T, 1, Eigen::Dynamic> a(size);
-  for (int i = 0; i < size; ++i) {
-    int n = rvalue_at(i, idx.head_);
-    math::check_range("row_vector[multi] indexing", name, rv.size(), n);
-    a(i) = rv.coeff(n - 1);
-  }
-  return a;
+template <typename EigMat, require_eigen_matrix_dynamic_t<EigMat>* = nullptr>
+inline auto rvalue(
+    EigMat&& a,
+    const cons_index_list<index_omni,
+                          cons_index_list<index_uni, nil_index_list>>& idx,
+    const char* name = "ANON", int depth = 0) {
+  math::check_range("matrix[uni] indexing", name, a.cols(), idx.tail_.head_.n_);
+  return a.col(idx.tail_.head_.n_ - 1).eval();
 }
 
 /**
- * Return the result of indexing the specified Eigen matrix with a
- * sequence consisting of one single index, returning a row vector.
+ * Return the result of indexing the specified Eigen vector with a
+ * sequence containing one single index, returning a scalar.
  *
- * Types:  mat[single] : rowvec
+ * Types:  vec[single] : scal
  *
- * @tparam T Scalar type.
- * @param[in] a Eigen matrix.
- * @param[in] idx Index consisting of one uni-index.
+ * @tparam EigVec An eigen vector
+ * @param[in] v Vector being indexed.
+ * @param[in] idx One single index.
  * @param[in] name String form of expression being evaluated.
  * @param[in] depth Depth of indexing dimension.
- * @return Result of indexing matrix.
+ * @return Result of indexing vector.
  */
-template <typename T>
-inline Eigen::Matrix<T, 1, Eigen::Dynamic> rvalue(
-    const Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>& a,
-    const cons_index_list<index_uni, nil_index_list>& idx,
-    const char* name = "ANON", int depth = 0) {
-  int n = idx.head_.n_;
-  math::check_range("matrix[uni] indexing", name, a.rows(), n);
-  return a.row(n - 1);
+template <typename EigVec, require_eigen_vector_t<EigVec>* = nullptr>
+inline auto&& rvalue(EigVec&& v,
```

Comment:
```suggestion
inline value_type_t<EigVec> rvalue(EigVec&& v,
```
If this is passed in a temporary it can return a dangling reference.